### ISSUE TITLE
Karpenter schedules pods in the same zone when using topologySpreadConstraint

### ISSUE THREAD

**Issue by nicolajknudsen** [aws/karpenter-provider-aws#5597]
### Description

**Observed Behavior**:
When using this config:
```yaml
  topologySpreadConstraints:
    - labelSelector:
        matchLabels:
          app.kubernetes.io/instance: grafana-agent-statefulset
          name: grafana-agent
      maxSkew: 1
      topologyKey: topology.kubernetes.io/zone
      whenUnsatisfiable: DoNotSchedule
```
to deploy Grafana Agent as a statefulSet the pods all gets scheduled in the same availability zone. 

**Expected Behavior**:
For the pods to be scheduled in different availability zones

**Reproduction Steps** (Please include YAML):
This is my full values file for the Grafana Agent helm chart (https://github.com/grafana/helm-charts/releases/download/grafana-agent-0.31.1/grafana-agent-0.31.1.tgz) :
```yaml
agent:
  mode: "flow"
  configMap:
    create: true
  clustering:
    enabled: true
  listenPort: 8080
  enableReporting: false
  resources:
    limits:
      cpu: 100m
      memory: 1200Mi
    requests:
      cpu: 50m
      memory: 600Mi

controller:
  type: "statefulset"
  replicas: 3
  volumeClaimTemplates:
    - apiVersion: v1
      kind: PersistentVolumeClaim
      metadata:
        name: storage
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 10Gi
        volumeMode: Filesystem
  tolerations:
    - key: "spot"
      operator: "Exists"
      effect: "NoSchedule"
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector:
            matchExpressions:
              - key: app.kubernetes.io/instance
                operator: In
                values:
                  - grafana-agent-statefulset
          topologyKey: "kubernetes.io/hostname"
    nodeAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - weight: 10
          preference:
            matchExpressions:
              - key: "karpenter.sh/capacity-type"
                operator: "In"
                values: ["spot"]
        - weight: 1
          preference:
            matchExpressions:
              - key: "karpenter.sh/capacity-type"
                operator: "In"
                values: ["on-demand"]
  topologySpreadConstraints:
    - labelSelector:
        matchLabels:
          app.kubernetes.io/instance: grafana-agent-statefulset
          name: grafana-agent
      maxSkew: 1
      topologyKey: topology.kubernetes.io/zone
      whenUnsatisfiable: DoNotSchedule
```
There are karpenter created nodes in three different AZ, but the three pods all gets scheduled in eu-west-1a. It even schedules one of them in the less preferred capacity type "on-demand".

I feel like I'm missing something really obvious here, please excuse me if this is not the right place to report this.

**Versions**:
- Chart Version:
Karpenter 0.31.3

- Kubernetes Version (`kubectl version`):
```yaml
serverVersion:
  buildDate: "2024-01-02T20:34:46Z"
  compiler: gc
  gitCommit: e78a4be9da4c375a87a109e0f4a5f4a8d2bc17c0
  gitTreeState: clean
  gitVersion: v1.28.5-eks-5e0fdde
  goVersion: go1.20.12
  major: "1"
  minor: 28+
  platform: linux/amd64
```

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by tzneal** [comment]
Can you provide the full pod spec for one of the stateful set pods?

**Comment by nicolajknudsen** [comment]
The output of `kubectl get pod grafana-agent-statefulset-0 -o yaml`

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubectl.kubernetes.io/default-container: grafana-agent
    kubectl.kubernetes.io/restartedAt: "2024-02-06T13:56:19+01:00"
  creationTimestamp: "2024-02-06T12:57:17Z"
  generateName: grafana-agent-statefulset-
  labels:
    app.kubernetes.io/instance: grafana-agent-statefulset
    app.kubernetes.io/name: grafana-agent
    apps.kubernetes.io/pod-index: "0"
    controller-revision-hash: grafana-agent-statefulset-5f8f6f548d
    statefulset.kubernetes.io/pod-name: grafana-agent-statefulset-0
  name: grafana-agent-statefulset-0
  namespace: monitoring
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: StatefulSet
    name: grafana-agent-statefulset
    uid: 192d2d25-54d1-4e88-8a28-36aa612653c1
  resourceVersion: "103252892"
  uid: f9037499-3f4e-483b-a75a-4db31717f337
spec:
  affinity:
    nodeAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - preference:
          matchExpressions:
          - key: karpenter.sh/capacity-type
            operator: In
            values:
            - spot
        weight: 10
      - preference:
          matchExpressions:
          - key: karpenter.sh/capacity-type
            operator: In
            values:
            - on-demand
        weight: 1
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: app.kubernetes.io/instance
            operator: In
            values:
            - grafana-agent-statefulset
        topologyKey: kubernetes.io/hostname
  containers:
  - args:
    - run
    - /etc/agent/config.river
    - --storage.path=/tmp/agent
    - --server.http.listen-addr=0.0.0.0:8080
    - --server.http.ui-path-prefix=/
    - --disable-reporting
    - --cluster.enabled=true
    - --cluster.join-addresses=grafana-agent-statefulset-cluster
    env:
    - name: AGENT_MODE
      value: flow
    - name: AGENT_DEPLOY_MODE
      value: helm
    - name: HOSTNAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.nodeName
    image: docker.io/grafana/agent:v0.39.1
    imagePullPolicy: IfNotPresent
    name: grafana-agent
    ports:
    - containerPort: 8080
      name: http-metrics
      protocol: TCP
    readinessProbe:
      failureThreshold: 3
      httpGet:
        path: /-/ready
        port: 8080
        scheme: HTTP
      initialDelaySeconds: 10
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 1
    resources:
      limits:
        cpu: 100m
        memory: 1200Mi
      requests:
        cpu: 50m
        memory: 600Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/agent
      name: config
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-kvpxh
      readOnly: true
  - args:
    - --volume-dir=/etc/agent
    - --webhook-url=http://localhost:8080/-/reload
    image: ghcr.io/jimmidyson/configmap-reload:v0.12.0
    imagePullPolicy: IfNotPresent
    name: config-reloader
    resources:
      requests:
        cpu: 1m
        memory: 5Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/agent
      name: config
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-kvpxh
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  hostname: grafana-agent-statefulset-0
  nodeName: ip-10-7-2-61.eu-west-1.compute.internal
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: grafana-agent-statefulset
  serviceAccountName: grafana-agent-statefulset
  subdomain: grafana-agent-statefulset
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoSchedule
    key: spot
    operator: Exists
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  topologySpreadConstraints:
  - labelSelector:
      matchLabels:
        app.kubernetes.io/instance: grafana-agent-statefulset
        app.kubernetes.io/name: grafana-agent
        name: grafana-agent
    matchLabelKeys:
    - app
    - pod-template-hash
    maxSkew: 1
    topologyKey: topology.kubernetes.io/zone
    whenUnsatisfiable: DoNotSchedule
  volumes:
  - name: storage
    persistentVolumeClaim:
      claimName: storage-grafana-agent-statefulset-0
  - configMap:
      defaultMode: 420
      name: grafana-agent-statefulset
    name: config
  - name: kube-api-access-kvpxh
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2024-02-06T12:57:17Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2024-02-06T12:57:27Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2024-02-06T12:57:27Z"
    status: "True"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2024-02-06T12:57:17Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: containerd://4df07ab4084782349a573cc4838e505665213e1efa7e3eded292527e440fc95f
    image: ghcr.io/jimmidyson/configmap-reload:v0.12.0
    imageID: ghcr.io/jimmidyson/configmap-reload@sha256:a7c754986900e41fc47656bdc8dfce33227112a7cce547e0d9ef5d279f4f8e99
    lastState: {}
    name: config-reloader
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2024-02-06T12:57:17Z"
  - containerID: containerd://ba3a0e3d80365fbdb7df5bc4baf9f65786e0c36d6c0a18ca01c7a449c3487ffb
    image: docker.io/grafana/agent:v0.39.1
    imageID: docker.io/grafana/agent@sha256:e19f3f0af97884affac3ee49012cc0163aed521d0475c95a29e037dccbcd92b3
    lastState: {}
    name: grafana-agent
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2024-02-06T12:57:17Z"
  hostIP: 10.7.2.61
  phase: Running
  podIP: 10.7.13.251
  podIPs:
  - ip: 10.7.13.251
  qosClass: Burstable
  startTime: "2024-02-06T12:57:17Z"
```

I have added the `matchLabelKeys` following some answer on SO on a similar issue to no effect. I've also installed the kubernetes descheduler in the cluster, but to no effect. 

One of the karpenter nodes in zone eu-west-1a got replaced earlier, but the replacement pod was spun up on the new node that is also placed in that same AZ.

The output of `kubectl describe pod` states that: "topology.kubernetes.io/zone:DoNotSchedule when max skew 1 is exceeded for selector app.kubernetes.io/instance=grafana-agent-statefulset,app.kubernetes.io/name=grafana-agent,name=grafana-agent" which is certainly true, but I don't get why though.

**Comment by nicolajknudsen** [comment]
Output sample from `kubectl logs -f -n karpenter -l app.kubernetes.io/name=karpenter -c controller`:

```log
2024-02-09T10:18:10.521Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1b]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/mimir-ingester-zone-b-0"}
2024-02-09T10:18:10.924Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":10,"preference":{"matchExpressions":[{"key":"karpenter.sh/capacity-type","operator":"In","values":["spot"]}]}}	{"commit": "74bf29c", "pod": "monitoring/mimir-gateway-d84dcd5cd-s4jr2"}
2024-02-09T10:18:10.929Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":100,"podAffinityTerm":{"labelSelector":{"matchLabels":{"app.kubernetes.io/component":"gateway","app.kubernetes.io/instance":"loki","app.kubernetes.io/name":"loki-distributed"}},"topologyKey":"failure-domain.beta.kubernetes.io/zone"}}	{"commit": "74bf29c", "pod": "monitoring/loki-loki-distributed-gateway-8dc69c8d-mbbwm"}
2024-02-09T10:18:11.016Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":100,"podAffinityTerm":{"labelSelector":{"matchLabels":{"app.kubernetes.io/component":"distributor","app.kubernetes.io/instance":"loki","app.kubernetes.io/name":"loki-distributed"}},"topologyKey":"failure-domain.beta.kubernetes.io/zone"}}	{"commit": "74bf29c", "pod": "monitoring/loki-loki-distributed-distributor-5c688b4745-vp7dw"}
2024-02-09T10:18:11.021Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":100,"podAffinityTerm":{"labelSelector":{"matchLabels":{"app.kubernetes.io/component":"querier","app.kubernetes.io/instance":"loki","app.kubernetes.io/name":"loki-distributed"}},"topologyKey":"failure-domain.beta.kubernetes.io/zone"}}	{"commit": "74bf29c", "pod": "monitoring/loki-loki-distributed-querier-0"}
2024-02-09T10:18:11.025Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":1,"preference":{"matchExpressions":[{"key":"karpenter.sh/capacity-type","operator":"In","values":["on-demand"]}]}}	{"commit": "74bf29c", "pod": "monitoring/mimir-gateway-d84dcd5cd-s4jr2"}
2024-02-09T10:18:21.631Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-1"}
2024-02-09T10:18:21.631Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-legacy-scraper-0"}
2024-02-09T10:18:21.632Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-0"}
2024-02-09T10:18:23.425Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-1"}
2024-02-09T10:18:23.426Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-legacy-scraper-0"}
2024-02-09T10:18:23.927Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-1"}
2024-02-09T10:18:25.039Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-legacy-scraper-0"}
2024-02-09T10:18:26.415Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/mimir-ingester-zone-a-0"}
2024-02-09T10:18:26.415Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/mimir-store-gateway-zone-a-0"}
2024-02-09T10:18:26.415Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-2"}
2024-02-09T10:18:27.515Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-0"}
2024-02-09T10:18:28.022Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1c]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/mimir-ingester-zone-c-0"}
2024-02-09T10:18:28.022Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1c]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/mimir-store-gateway-zone-c-0"}
2024-02-09T10:18:28.215Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":10,"preference":{"matchExpressions":[{"key":"karpenter.sh/capacity-type","operator":"In","values":["spot"]}]}}	{"commit": "74bf29c", "pod": "monitoring/mimir-gateway-d84dcd5cd-vtnnx"}
2024-02-09T10:18:28.217Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":10,"preference":{"matchExpressions":[{"key":"karpenter.sh/capacity-type","operator":"In","values":["spot"]}]}}	{"commit": "74bf29c", "pod": "monitoring/mimir-gateway-d84dcd5cd-7qkqk"}
2024-02-09T10:18:28.223Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":1,"preference":{"matchExpressions":[{"key":"karpenter.sh/capacity-type","operator":"In","values":["on-demand"]}]}}	{"commit": "74bf29c", "pod": "monitoring/mimir-gateway-d84dcd5cd-vtnnx"}
2024-02-09T10:18:28.314Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":1,"preference":{"matchExpressions":[{"key":"karpenter.sh/capacity-type","operator":"In","values":["on-demand"]}]}}	{"commit": "74bf29c", "pod": "monitoring/mimir-gateway-d84dcd5cd-7qkqk"}
2024-02-09T10:18:29.033Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1b]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/mimir-ingester-zone-b-0"}
2024-02-09T10:18:29.816Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":10,"preference":{"matchExpressions":[{"key":"karpenter.sh/capacity-type","operator":"In","values":["spot"]}]}}	{"commit": "74bf29c", "pod": "monitoring/mimir-gateway-d84dcd5cd-s4jr2"}
2024-02-09T10:18:29.819Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":100,"podAffinityTerm":{"labelSelector":{"matchLabels":{"app.kubernetes.io/component":"gateway","app.kubernetes.io/instance":"loki","app.kubernetes.io/name":"loki-distributed"}},"topologyKey":"failure-domain.beta.kubernetes.io/zone"}}	{"commit": "74bf29c", "pod": "monitoring/loki-loki-distributed-gateway-8dc69c8d-mbbwm"}
2024-02-09T10:18:29.821Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":100,"podAffinityTerm":{"labelSelector":{"matchLabels":{"app.kubernetes.io/component":"distributor","app.kubernetes.io/instance":"loki","app.kubernetes.io/name":"loki-distributed"}},"topologyKey":"failure-domain.beta.kubernetes.io/zone"}}	{"commit": "74bf29c", "pod": "monitoring/loki-loki-distributed-distributor-5c688b4745-vp7dw"}
2024-02-09T10:18:29.825Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":100,"podAffinityTerm":{"labelSelector":{"matchLabels":{"app.kubernetes.io/component":"querier","app.kubernetes.io/instance":"loki","app.kubernetes.io/name":"loki-distributed"}},"topologyKey":"failure-domain.beta.kubernetes.io/zone"}}	{"commit": "74bf29c", "pod": "monitoring/loki-loki-distributed-querier-0"}
2024-02-09T10:18:29.827Z	DEBUG	controller.deprovisioning	relaxing soft constraints for pod since it previously failed to schedule, removing: spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[0]={"weight":1,"preference":{"matchExpressions":[{"key":"karpenter.sh/capacity-type","operator":"In","values":["on-demand"]}]}}	{"commit": "74bf29c", "pod": "monitoring/mimir-gateway-d84dcd5cd-s4jr2"}
2024-02-09T10:18:41.426Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-1"}
2024-02-09T10:18:41.426Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-legacy-scraper-0"}
2024-02-09T10:18:41.426Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-0"}
2024-02-09T10:18:42.234Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-1"}
2024-02-09T10:18:42.234Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-legacy-scraper-0"}
2024-02-09T10:18:43.626Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-1"}
2024-02-09T10:18:44.843Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-legacy-scraper-0"}
2024-02-09T10:18:45.630Z	DEBUG	controller.deprovisioning	adding requirements derived from pod volumes, [{topology.kubernetes.io/zone In [eu-west-1a]} {topology.kubernetes.io/region In [eu-west-1]}]	{"commit": "74bf29c", "pod": "monitoring/grafana-agent-statefulset-0"}
```

**Comment by nicolajknudsen** [comment]
I figured it out, there is no bug here, just a lack of knowledge on my part. 

First of all the label selector is wrong, it should be:
```yaml
topologySpreadConstraints:
    - labelSelector:
        matchExpressions:
          - key: app.kubernetes.io/instance
            operator: In
            values:
              - grafana-agent-statefulset
      maxSkew: 1
      topologyKey: topology.kubernetes.io/zone
      whenUnsatisfiable: DoNotSchedule
```
but more importantly, the application is deployed as a statefulSet with attached PVC's, and since it was already running before adding the topologySpreadConstraint, all those PVC's were already bound (originally arbitrarily I assume) to the eu-west-1a AZ.

These can of course not be moved around by the scheduler.

The combination of these two issues made the logs look like everything was ok. Once the labelSelector got fixed, the nodes became unschedulable and the underlying problem became apparent.

My solution was to manually delete the statefulSet, and each of the PVC's. This might not be a solution for everyone in a production setting. I think there is a way to migrate the volume data, but I didn't look into that.

Thank you for your time reviewing this.
